sample/validator-test/Main.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{- | Module: Main
+ Copyright: (C) Liqwid Labs 2022
+ License: Apache 2.0
+ Portability: GHC only
+ Stability: Experimental
+ Example of @plutarch-quickcheck@ tests. These are meant to be read as source
+ code.
+-}
+module Main (main) where
+
+import GHC.IO.Encoding (setLocaleEncoding, utf8)
+import Plutarch.Builtin (pforgetData)
+import Plutarch.Prelude (
+  PData,
+  PInteger,
+  POpaque,
+  S,
+  Term,
+  pconstant,
+  pdata,
+  pfromData,
+  pif,
+  plam,
+  popaque,
+  ptraceError,
+  ptryFrom,
+  tcont,
+  unTermCont,
+  (#),
+  (#<),
+  (:-->),
+ )
+import Plutarch.Test.QuickCheck (fromFailingPPartial, fromPPartial, pconstantT)
+import Test.Tasty (adjustOption, defaultMain, testGroup)
+import Test.Tasty.QuickCheck (
+  Property,
+  QuickCheckTests,
+  chooseInteger,
+  forAll,
+  testProperty,
+ )
+
+-- This is an example pseudo validator that will pass when the sum of parameter
+-- and the datum is less than 20.
+-- Example doesn't use actual 'PValidator' because 'PScriptContext' is hard to
+-- generate without PCB(Plutarch Context Builder).
+myValidator ::
+  forall (s :: S).
+  Term s (PInteger :--> PData :--> PData :--> POpaque)
+myValidator = plam $ \param dat _red -> unTermCont $ do
+  (dat', _) <- tcont $ ptryFrom dat
+
+  pure $
+    pif
+      (pfromData dat' + param #< 20)
+      (popaque $ pconstant ())
+      (ptraceError "param + datum is larger than 20")
+
+-- Property test. This uses 'fromPPartial'; it will fail when the validator
+-- fails and will pass when validator succeed.
+valProp :: Property
+valProp =
+  forAll (pconstantT <$> chooseInteger (0, 14)) $
+    fromPPartial valScript
+  where
+    valScript :: forall (s :: S). Term s (PInteger :--> POpaque)
+    valScript = plam $ \x ->
+      myValidator
+        # 5
+        # pforgetData (pdata x)
+        # pforgetData (pdata x)
+
+valPropFail :: Property
```

Comment:
A comment on how does it differ from the one above would be nice here 🙏 (i.e. why it uses the failing function instead.